docs/proposals/initializers.md

```diff
@@ -0,0 +1,110 @@
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/resource-qos.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Initializers in Kubernetes
+
+**Author**: Derek Carr (@derekwaynecarr)
+
+**Status**: Proposed
+
+*This document introduces the concept of initializers in Kubernetes, and describes use cases and implementation details.*
+
+## Motivation
+
+The ability to introduce hooks into the API to modify an object prior to it being acted upon in the system is a common requirement.
+
+The primary mechanism to achieve this use-case today is to write an "admission controller" which is a server-side filter that observes API server actions, and provides the ability for an ordered set of filters to observe, modify, accept, or reject an object from being persisted into the API server.  By providing the ability to give an immediate yes/no response to the user, it provides a good place for hooks for use-cases that require an immediate user response.  For example, the user is able to be told that they are not able to create an object in a namespace without first creating that namespace.  Another common example is quota enforcement where the user is immediately forbidden from exceeding any quota constraints.  In essence, its a synchronous visitor pattern.
+
+While powerful, admission controllers have drawbacks.  The major drawback is that it requires a recompilation and restart of the API server to introduce new behavior.  For those scenarios where an immediate yes/no response was not required for the user experience, it would be ideal if the system could support an asynchronous visitor pattern that would allow controllers external to the API server to observe, and modify a resource prior to it becoming 'active' in the system.  This would allow a new type of controller to be authored in the system that can initialize a resource similar to how other controllers watch and act upon already initialized objects.  There are many scenarios where this would be appropriate.  For example, a pod auto-sizer could be deployed separate from the API server that watches for incoming pods.  The pod-autosizer should be able to assign compute resource requests prior to the pod being scheduled.  If authored as an "initializer", it would allow anyone to bring their own logic to the cluster for how sizing occurs without needing to recompile or change the core components.
```

Comment:
The pod autosizer probably wasn't a great example because we already have this, as a synchronous admission controller: https://github.com/kubernetes/kubernetes/blob/master/docs/admin/admission-controllers.md#initialresources-experimental

The plan for making it customizable is that it will fetch predictions via an API, so you replace the component that provides the predictions. (Today the prediction logic is compiled into the admission controller.)